designs/v1-api.md

```diff
@@ -0,0 +1,529 @@
+2. Stability: The change ensures proper operational excellence for behavior that is leaky or has race conditions in the beta state
+3. Planned Deprecations: The change cleans-up deprecations that were previously planned the project
+
+## Migration Path
+
+Karpenter will **not** be changing its API group or resource kind as part of the v1 API bump. By avoiding this, we can leverage the [existing Kubernetes conversion webhook process](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion) for upgrading APIs, which allows upgrades to newer versions of the API to occur in-place, without any customer intervention or node rolling. The upgrade process will be executed as follows:
+
+1. Apply the updated NodePool, NodeClaim, and EC2NodeClass CRDs, which will contain a `v1` versions listed under the `versions` section of the CustomResourceDefinition
+2. Upgrade Karpenter controller to its `v1` version. This version of Karpenter will start reasoning in terms of the `v1` API schema in its API requests. Resources will be converted from the v1beta1 to the v1 version at runtime, using conversion webhooks shipped by the upstream Karpenter project and the Cloud Providers (for NodeClass changes).
+3. Users update their `v1beta1` manifests that they are applying through IaC or GitOps to use the new `v1` version.
+4. Karpenter drops the `v1beta1` version from the `CustomResourceDefinition` and the conversion webhooks on the next minor version release of Karpenter, leaving no webhooks and only the `v1` version still present
+
+## NodePool API
+
+```
+apiVersion: karpenter.sh/v1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  template:
+    metadata:
+      labels:
+        billing-team: my-team
+      annotations:
+        example.com/owner: "my-team"
+    spec:
+      nodeClassRef:
+        group: karpenter.k8s.aws # Updated since only a single version will be served
+        kind: EC2NodeClass
+        name: default
+      taints:
+        - key: example.com/special-taint
+          effect: NoSchedule
+      startupTaints:
+        - key: example.com/another-taint
+          effect: NoSchedule
+      requirements:
+        - key: "karpenter.k8s.aws/instance-category"
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2 # Alpha field, added for spot best practices support
+  disruption:
+    budgets:
+      - nodes: 0
+        schedule: "0 9 * * mon-fri"
+        duration: 8h
+      - nodes: "10%"
+    expireAfter: 720h | Never
+    consolidationPolicy: WhenUnderutilized | WhenEmpty
+    consolidateAfter: 1m | Never # Added to allow additional control over consolidation aggressiveness
+  weight: 10
+  limits:
+    cpu: "1000"
+    memory: 1000Gi
+status:
+  conditions:
+    # Abnormal True Conditions
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: NodeClassReady
```

Comment:
Is this an abnormal-true condition? I'd expect `false` to indicate abnormality here.

Reply:
Nope, mistype. I'll just drop this from the example.